synapse/static/client/register/js/register.js

```diff
@@ -1,67 +1,93 @@
+                for (j=0; j<flow.stages.length; j++) {
+                    var stage = flow.stages[j];
+                    if ("m.login.recaptcha" === stage) {
+                        serverExpectsCaptcha = true;
+                        break;
+                    }
+                }
+            }
+            if (!serverExpectsCaptcha) {
+                console.log("This server does not require a captcha.");
+                return;
+            }
+            console.log("Setting up ReCaptcha for "+matrixRegistration.endpoint);
+            var public_key = response.params["m.login.recaptcha"].public_key;
+            if (public_key === undefined) {
+                console.error("No public key defined for captcha!");
+                setFeedbackString("Misconfigured captcha for server. Contact server admin.");
+                return;
+            }
+            Recaptcha.create(public_key,
+            "regcaptcha",
+            {
+                theme: "red",
+                callback: Recaptcha.focus_response_field
+            });
+            window.matrixRegistration.isUsingRecaptcha = true;
         }
-        if (!serverExpectsCaptcha) {
-            console.log("This server does not require a captcha.");
-            return;
-        }
-        console.log("Setting up ReCaptcha for "+matrixRegistration.endpoint);
-        var public_key = window.matrixRegistrationConfig.recaptcha_public_key;
-        if (public_key === undefined) {
-            console.error("No public key defined for captcha!");
-            setFeedbackString("Misconfigured captcha for server. Contact server admin.");
-            return;
-        }
-        Recaptcha.create(public_key,
-        "regcaptcha",
-        {
-            theme: "red",
-            callback: Recaptcha.focus_response_field
-        });
-        window.matrixRegistration.isUsingRecaptcha = true;
-    }).error(errorFunc);
-    
+    })
 };
 
 var submitCaptcha = function(user, pwd) {
     var challengeToken = Recaptcha.get_challenge();
     var captchaEntry = Recaptcha.get_response();
     var data = {
-        type: "m.login.recaptcha",
-        challenge: challengeToken,
-        response: captchaEntry
+        auth: {
+            type: "m.login.recaptcha",
+            challenge: challengeToken,
+            response: captchaEntry,
+            session: window.matrixRegistration.session,
+        },
+        username: user,
+        password: pwd,
     };
     console.log("Submitting captcha");
-    $.post(matrixRegistration.endpoint, JSON.stringify(data), function(response) {
-        console.log("Success -> "+JSON.stringify(response));
-        submitPassword(user, pwd, response.session);
-    }).error(function(err) {
-        Recaptcha.reload();
-        errorFunc(err);
-    });
+    $.ajax({
+        url: matrixRegistration.endpoint,
+        method: "POST",
+        data: JSON.stringify(data),
+        complete: function(response) {
```

Comment:
As a general rule, I prefer the promise syntax (`$.ajax({...}).done( f )`) over the callback syntax (`$.ajax({success: f})`). I find it easier to read, and has a lower tendency to lead to lots of indentation. In fact, all the reasons promises were introduced to JS.

In this case, separate `done` and `error` handlers would be clearer than a combined one which then has to go and figure out if we had success or fail.